osquery/tables/system/linux/portage.cpp

```diff
@@ -0,0 +1,416 @@
+ */
+
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+#include <boost/algorithm/string.hpp>
+#include <boost/algorithm/string/join.hpp>
+#include <boost/algorithm/string/predicate.hpp>
+#include <boost/algorithm/string/trim_all.hpp>
+#include <boost/concept_check.hpp>
+#include <boost/filesystem/operations.hpp>
+#include <boost/filesystem/path.hpp>
+
+#include <osquery/core.h>
+#include <osquery/filesystem.h>
+#include <osquery/logger.h>
+#include <osquery/tables.h>
+
+#include "osquery/core/conversions.h"
+
+#define fs boost::filesystem
+
+namespace osquery {
+namespace tables {
+
+/* File paths used by portage */
+fs::path portagePackageDir = "/var/db/pkg";
+fs::path portageKeywords = "/etc/portage/package.keywords";
+fs::path portageEtcMake = "/etc/make.conf";
+fs::path portageEtcPortageMake = "/etc/portage/make.conf";
+fs::path portageMask = "/etc/portage/package.mask";
+fs::path portageUnMask = "/etc/portage/package.unmask";
+fs::path portageUse = "/etc/portage/package.use";
+fs::path portageWorld = "/var/lib/portage/world";
+
+/* Shared functions and structs */
+
+/**
+ * @brief container to hold data before we store it to QueryData
+ */
+struct portagePackageHolder {
+ public:
+  std::string package;
+  std::string version;
+  std::string keyword;
+  std::string mask;
+  std::string unmask;
+};
+
+/**
+ * @brief split a package string with version into package name and package
+ * version.
+ *
+ * we need to split a package stiring which includes the version it affects into
+ * a pair of package name (first) and package version (second).
+ */
+std::pair<std::string, std::string> portageSplitPackageVersion(
+    std::string pkgStr) {
+  std::pair<std::string, std::string> package = std::make_pair("", "");
+  while (boost::starts_with(pkgStr, ">") || boost::starts_with(pkgStr, "<") ||
+         boost::starts_with(pkgStr, "=")) {
+    if (!(package.second.length() == 0 && boost::starts_with(pkgStr, "="))) {
+      package.second += pkgStr.substr(0, 1);
+    }
+    pkgStr.erase(0, 1);
+  }
+  auto dpos = pkgStr.find('/');
+  auto vpos = pkgStr.find_last_of('-');
+  if (vpos != std::string::npos && vpos > dpos && vpos < pkgStr.length() - 1 &&
+      isdigit(pkgStr[vpos + 1])) {
+    package.second += pkgStr.substr(vpos + 1);
+    pkgStr.erase(vpos, pkgStr.length() - vpos);
+  } else {
+    vpos = pkgStr.find_last_of('-', vpos - 1);
+    if (vpos != std::string::npos && vpos > dpos &&
+        vpos < pkgStr.length() - 1 && isdigit(pkgStr[vpos + 1])) {
+      package.second += pkgStr.substr(vpos + 1);
+      pkgStr.erase(vpos, pkgStr.length() - vpos);
+    }
+  }
+  package.first = pkgStr;
+  return package;
+}
+
+/* Functions to fetch content for database tables */
+QueryData parsePortagePackages(const fs::path& pkgpath,
+                               const std::string& worldContent) {
+  QueryData results;
+
+  fs::path full_path(fs::initial_path());
+  full_path = fs::system_complete(fs::path(pkgpath));
```

Comment:
Be very careful with `boost::filesystem` API calls! The simple variant of each will throw exceptions! Either `try` and handle the exception for each or use the `boost::system::error_code` variant.